### DISCUSSION TITLE
Detecting Run Finish

### DISCUSSION THREAD

**Discussion by agoldis** [sorry-cypress/sorry-cypress#287] [Ideas]
@DeniDoman @tico24 @fsmaia thought you'd be interested

# TL;DR Where to run the timers?

In order to support `RUN_FINISH` implementation (see below), we need to implement timers that trigger the logic behind `RUN_FINISH`. This logic is implemented in director service (see below)

The director service is supposed to:

- ~to be stateless - i.e. run on lambda etc~ edit: seems like that would be only supported on cloud platforms with a proper always-on notification services
- to support concurrent setup (i.e. more than 1 instances of the director)
- to be crash-prone - if docker crashes, a new instance should be able to resume a run

Having the timers in-memory does not support those. However, it's the simplest option in terms of implementation. 

An alternative is using an external service that fires scheduled events and "wakes up" the director when time comes. 

<img width="833" alt="Screen Shot 2021-02-24 at 1 21 19 AM" src="https://user-images.githubusercontent.com/1637928/108978644-aa2c1e00-763e-11eb-8d8a-e3fae976c061.png">

Some candidates for such external services:

- ~Redis~  edit: silly me, redis still requires a listener
- Postgres (with triggers mechanism) < really not sure about it
- Some sort of queue like RabbitMQ

--- 

The questions are:
- based on your usage do you prefer in-memory or an extra service for sorry-cypress?
- how difficult would it be to deploy an extra service?
- @tico24 @fsmaia would it be hard to change the charts accordingly?

# Detecting Run Finish

Run - is an aggregative abstraction of groups of specs

Group -  is a collection of specs

Spec - is a collection of tests

Instance - is a spec that was claimed and sent to cypress agent

RUN_START - an event emitted when we create a run, needs to be emitted once and only once for a run

INSTANCE_START - an event emitted when the machine claims a spec

INSTANCE_FINISH - an event emitted when machine reports result for an instance

RUN_FINISH - an event emitted when the run has finished, needs to be emitted once and only once

### Problem

When do we emit RUN_FINISH

### Naive approach

Wait for all specs to finish. Doesn't work because:

- **Problem A** An instance can never report back as finished (e.g. cypress agent crashes, user cancels CI build)
- **Problem B** Different groups can have different specs. We don't always know all the specs in advance, thus we can emit `RUN_FINISH` prematurely before we know all the specs.

E.g. given two groups of specs and 2 machines:

```
groupA
 - spec A1

groupB
- spec B1
```

1. machine 1 runs groupA, creates a run, populates specs for group A, gets spec A1
2. machine 1 finishes specA1 after 30 seconds
3. 💥 emitting RUN_FINISH because all known specs are completed
4. machine 2 joins the run after 60 seconds, adds new specs for group B, gets spec B1
5. machine 2 finishes specB1 after 30 seconds
6. 💥emitting RUN_FINISH because all known specs are completed

<img width="1258" alt="Screen_Shot_2021-02-24_at_12 43 08_AM" src="https://user-images.githubusercontent.com/1637928/108978285-4d306800-763e-11eb-973d-d220517d8d74.png">

## Option A

Have a preset timeout for each run. If the run doesn't finish within `RUN_TIMEOUT` (let's say an hour) emit RUN_FINISH.

**Problems**

Does solve the problem B - we can prematurely emit `RUN_FINISH`

Long wait time for CI - e.g. GH PR might wait for up to `RUN_TIMEOUT`

## Option B

I think that's the [original cypress dashboard](https://docs.cypress.io/guides/guides/parallelization.html#Run-completion-delay) solution. We wait for up to `RUN_COMPLETION_DELAY` after last instance **finished.** When `RUN_COMPLETION_DELAY` fires we emit `RUN_FINISH`

**Problems**

- There's always a delay of `RUN_COMPLETION_DELAY` between when run actually finishes and when `RUN_FINISH` emitted
- If we start the timer with `RUN_START`, then the very first instance has to finish within `RUN_COMPLETION_DELAY`

<img width="1459" alt="Screen_Shot_2021-02-24_at_12 29 15_AM" src="https://user-images.githubusercontent.com/1637928/108978325-53bedf80-763e-11eb-88f5-230feeae78d9.png">

![Detecting%20Run%20Finish%205503b217aaad4613a348eb2e9ab1d8f2/Screen_Shot_2021-02-24_at_12.29.15_AM.png](Detecting%20Run%20Finish%205503b217aaad4613a348eb2e9ab1d8f2/Screen_Shot_2021-02-24_at_12.29.15_AM.png)

- If we start the timer only after the first `INSTANCE_FINISH`, and if the first instances never finish, we never emit `RUN_FINISH`

<img width="1410" alt="Screen_Shot_2021-02-24_at_12 31 30_AM" src="https://user-images.githubusercontent.com/1637928/108978351-591c2a00-763e-11eb-8e09-f379625ec42c.png">

- For some edge cases, we can emit `RUN_FINISH` prematurely if an instance is running longer than `RUN_COMPLETION_DELAY` and no other instances finish meanwhile.

<img width="1261" alt="Screen_Shot_2021-02-24_at_12 39 16_AM" src="https://user-images.githubusercontent.com/1637928/108978384-60dbce80-763e-11eb-8b2b-68c249104cbb.png">

## Option C

We wait for up to `INACTIVITY_TIMEOUT` after the last instance **started.** When `INACTIVITY_TIMEOUT` fires we emit `RUN_FINISH`

<img width="1287" alt="Screen_Shot_2021-02-24_at_12 54 13_AM" src="https://user-images.githubusercontent.com/1637928/108978401-65a08280-763e-11eb-958b-36b036d41e19.png">

**Problems**

- There's some delay of `INACTIVITY_TIMEOUT - LAST_SPEC` between when the run actually finishes and when `RUN_FINISH` emitted. It is shorter than option B.
- We emit `RUN_FINISH` prematurely if the last instance runs longer than `INACTIVITY_TIMEOUT`

<img width="1286" alt="Screen_Shot_2021-02-24_at_1 01 12_AM" src="https://user-images.githubusercontent.com/1637928/108978420-6a653680-763e-11eb-8856-04ca6a5c7519.png">

--- 

- What is the preferred options (I picked A)
- Did I miss something?
- Are there any other options/solutions?

**Comment by tico24** [comment]
From a k8s perspective, I'd steer away from postgres.

I've used redis quite a bit and it seems fine, should be reasonably easy to add to the helm chart.

**Comment by fsmaia** [comment]
How about the following approach?

When the first instance starts, we trigger a continuous check for the active instances count, and when it comes to zero and the tests aren't complete yet, we wait until another instance starts for a `RUN_INACTIVITY_TIMEOUT`. If that doesn't happen, then we emit `RUN_FINISH`.

If we expose the active instances count and the run status (complete specs vs. total specs) in the Mongo database, then we can fulfill all of the director requirements:
>
> * to be stateless - i.e. run on lamda etc
> * to support concurrent setup (i.e. more than 1 instances of director)
> * to be crash-prone - if docker crashes, a new instance should be able to resume a run

**Reply by agoldis** [reply]
To make sure I understand
>  trigger a continuous check for the active instances count,

What does continuous mean?

**Reply by fsmaia** [reply]
Probably polling to the database.

**Reply by agoldis** [reply]
>  when it comes to zero ...

What if it never comes to zero? E.g. machine crashed and doesn't report its completion
Or never changes from zero? E.g. no machines claimed any test

**Reply by fsmaia** [reply]
How about polling/checking over the remaining specs?

Example: if a run has 5 specs remaining, then we can check if this number still the same after a configured timeout.

Maybe we should deal with two timeouts: instance (drop an instance if it hasn't reported completion or hasn't claimed any specs for a given timeout) and run (drop all instances and emit `RUN_FINISH` if the total remaining specs still the same)

Guess it is kinda similar to Option B.

**Reply by agoldis** [reply]
Yeah... I've started with option C already - was curious about the technicalities to implement timers, we can change it if people don't like how it works
